src/Microsoft.Azure.WebJobs.Host/Bindings/BindingDataProvider.cs

```diff
@@ -96,15 +96,37 @@ public static BindingDataProvider FromType(Type type)
             }
 
             Dictionary<string, Type> contract = new Dictionary<string, Type>(StringComparer.OrdinalIgnoreCase);
+            Dictionary<string, Type> declaringTypes = new Dictionary<string, Type>(StringComparer.OrdinalIgnoreCase);
+
             foreach (PropertyHelper property in bindingDataProperties)
             {
-                if (contract.ContainsKey(property.Name))
+                if (!contract.ContainsKey(property.Name))
+                {
+                    contract.Add(property.Name, property.Property.PropertyType);
+                    declaringTypes.Add(property.Name, property.Property.DeclaringType);
+                }
+                else
                 {
-                    throw new InvalidOperationException(
-                        string.Format(CultureInfo.InvariantCulture,
-                        "Multiple properties named '{0}' found in type '{1}'.", property.Name, type.Name));
+                    // If the mapped property is declared on the type being bound, then the property was
+                    // shadowed.  Consider the direct property to have priority and replace the existing member.
```

Comment:
You can have shadow properties all up the hierarchy though, not just the most derived, right? So it doesn't seem this solves the general case. Add a test case with 4 levels with shadowing at a couple levels to verify.

It seems what we want is to simply use the most derived property for the contract. The problem is the reflection APIs return separate properties for shadow properties that change the return Type. We could order the properties here based on place in the hierarchy (I don't think we can assume any ordering from Type.GetRuntimeProperties()). Then the caller can just implement first one wins.

Reply:
I think we want to use the property as long as it is directly on the type being bound to. If there is another shadowed property on a more derived type, I think we still want things to work, so we may need to relax the validation so that we only throw if there are multiple properties with the same name AND we didn't find the property on the bound type.

Reply:
Since these properties would be using "new", I don't think it is correct in general to use the most derived property.

Reply:
That's correct; it's an explicit non-goal of this implementation to support shadowed properties deeper in the hierarchy.  _(see: #2830)_  

The rationale for this is that we don't have a strong heuristic to apply in the general case to capture the caller's intent.   For this specific case, we're being explicitly asked to bind to `type`, which makes it reasonable to assume that its direct properties are preferred, and it is most likely the type that was serialized.   

Once we start going further down, can we reasonably assume that the intent is to bind to the most direct ancestor?  Maybe.  It would support the statement that `type` was most likely what was serialized. 

We could start explicitly walking the hierarchy and accepting the property on the most direct ancestor, but do we think this is a mainstream enough case to take the performance hit?  Should we put the engineering investment in to try and include an optimized path for the non-shadowed case and fall back to the more robust inspection if we do detect shadowing?  I don't know that this scenario is common enough for either; I'm not sure that I would advocate for a more robust approach until we can prove that we need it.  Worst case, we can always revisit and extend.

With respect to tests, `FromType_DuplicatePropertiesDetected_Throws` on line 57 verifies that shadowing in the hierarchy is not supported, following the existing behavior of throwing an `InvalidOperationException`.

Reply:
Deleted my comments as I wasn't thinking about this correctly and @jsquire summarized the considerations.

Reply:
For a trigger binding of type T, only T and derived types can be bound, not ancestors. So for a BindingDataProvider for T, it will only ever be asked to resolve properties of T or derived types.

The intent of the binding contract is to provide the value for each contract property from the instance being bound. I.e. what value would you get if you invoke each contract property by name on the instance being bound to. Those would be  the most derived property definitions.

As far performance, this isn't a concern since property helpers are created once then cached. I think something like I try in https://github.com/Azure/azure-webjobs-sdk/tree/bindingdata-fix would work.

Reply:
Yup; that's pretty similar to the approach that I was thinking of for "walk the hierarchy".   If we're not concerned with the performance impact, then we could bypass my thoughts on an optimized path and a fallback path and just pay for the additional allocations and reflection for all cases, which simplifies things.

I'm not 100% clear on the safety, and it may just be my ignorance, but I'm assuming someone chose to not support shadowing for a reason.   The case that I'm thinking of is something like:

```csharp
public class Top : SingleBase
{
    public string Name { get; set; }
}

public class SingleBase : BaseBase
{
    public int Count { get; set; }
}

public class BaseBase
{
    public long Count { get; set; }
}
```

Let's say that, for some reason, the service received a serialized instance of `BaseBase` and returns it to the trigger with the following:

```json
{
    "Count": 9223372036854775804
}
```

 The trigger attempts to bind to `Top` so when things get bound, we end up either:

1) Throwing when attempting to bind 

1)  With an instance that has an incorrect value due to the down-casting:

```csharp
instance of Top
{
    public string Name = <null>,
    public int Count = -4
}
```

To be transparent, I do not have enough depth of knowledge to know whether this is a legit concern or whether there are additional safeguards in there.   In the first case where we throw, assuming that gets surfaced correctly, then it seems like we're good.   In the second case or if that exception gets swallowed, that seems like we'd be doing the wrong thing.

From your perspective, do we need to be concerned there, or do we think that building a map of the visible properties of `type` by walking the hierarchy is safe?

Reply:
Yes you're right. Initially I was thinking that "new" properties completely replaced the base properties they hide, but further experimentation shows that they are still there and usable. E.g. for an instance of Top t, I can set both t.Count = int.MaxValue and ((BaseBase)t).Count = long.MaxValue.

Likely there was a reason why this hard exclusion was put in place, but I'm not aware of the histor. In the code there is actually a [PropertyHelper.GetVisibleProperties](https://github.com/Azure/azure-webjobs-sdk/blob/3f4ec78be9f43bb041937425ced00b341883aa42/src/Microsoft.Azure.WebJobs.Host/PropertyHelper.cs#L140) method which we're not using for binding providers.

Reply:
I think we can simply switch to using the GetVisibleProperties method. There's low risk for breaking changes here, since we don't currently support types with duplicate/hidden properties, and for types w/o such properties GetVisibleProperties returns the same results as GetProperties.

I prototyped this in https://github.com/Azure/azure-webjobs-sdk/tree/bindingdata-fix.

Reply:
Yes, that's what I'm thinking